### PR TITLE
Fix learning rate problems in run_classifier.py

### PR DESCRIPTION
**Pull request by matej-svejda**
- Don't do warmup twice (in BertAdam and manually)
- Compute num_train_steps correctly for the case where gradient_accumulation_steps > 1. The current version might lead the the LR never leaving the warmup phase, depending on the value of gradient_accumulation_steps.

With these changes I get > 84% accuracy on MRPC, without them its around 77%.

### CODE REVIEW

**Comment by thomwolf** [comment]
Thanks @matej-svejda.
So this problem was actually introduced by adding NVIDIA's fp16 optimizer (`FusedAdam`) to the examples. This optimizer is a simple Adam which doesn't incorporate a learning rate schedule so we had to add a manual learning rate schedule in the examples.
So a better solution is to keep the `warmup_linear` function but to only modify the learning rates when the fp16 optimiser is used (i.e. updating the weights only if `args.fp16==True`).
Also it would  be great to update the other examples similarly.
Do you want to do that in your PR?
I can also do that if you don't have the time.

**Comment by matej-svejda** [comment]
Sure, I can do that. Wanted to try out fp16 anyways :+1:

**Comment by matej-svejda** [comment]
@thomwolf Something like this?

File: examples/run_lm_finetuning.py [hunk 1]
```diff
@@ -527,7 +521,7 @@ def main():
         train_dataset = BERTDataset(args.train_file, tokenizer, seq_len=args.max_seq_length,
                                     corpus_lines=None, on_memory=args.on_memory)
         num_train_steps = int(
-            len(train_dataset) / args.train_batch_size / args.gradient_accumulation_steps * args.num_train_epochs)
+            len(train_dataset) / args.train_batch_size * args.num_train_epochs)
```
[the remark this answers is not in the document]
Comment (thomwolf) [reply examples/run_lm_finetuning.py]: Actually `num_train_steps` should also be divided by `torch.distributed.get_world_size()` if `args.local_rank != -1` so that this works well in distributed settings. This was fixed in `run_squad.py` (see around line 825 of `run_squad.py`). Could you update it here also?

[the remark this answers is not in the document]
Comment (thomwolf) [reply examples/run_lm_finetuning.py]: `run_classifier.py` should have the same general formula, do you think you can put them all on the same page since you are fixing this? 😅 Sorry it falls on you to harmonize all these examples with the best practices.

So the general formula should be:
```python
num_train_steps =
            len(train_dataset) // args.train_batch_size // args.gradient_accumulation_steps * args.num_train_epochs
if args.local_rank != -1:
    num_train_steps = num_train_steps // torch.distributed.get_world_size()
```

File: examples/run_squad2.py [hunk 2]
```diff
@@ -887,7 +882,7 @@ def main():
         train_examples = read_squad_examples(
             input_file=args.train_file, is_training=True)
         num_train_steps = int(
-            len(train_examples) / args.train_batch_size / args.gradient_accumulation_steps * args.num_train_epochs)
```
Comment (thomwolf) [inline examples/run_squad2.py]: Same here also

File: examples/run_swag.py [hunk 3]
```diff
@@ -358,7 +353,7 @@ def main():
     if args.do_train:
         train_examples = read_swag_examples(os.path.join(args.data_dir, 'train.csv'), is_training = True)
         num_train_steps = int(
-            len(train_examples) / args.train_batch_size / args.gradient_accumulation_steps * args.num_train_epochs)
+            len(train_examples) / args.train_batch_size * args.num_train_epochs)
```
Comment (thomwolf) [inline examples/run_swag.py]: And here

**Review by thomwolf — CHANGES_REQUESTED** [review]
Ok went through it. It looks great. I just have a remark on the total number of training steps that you can see in the comments.

File: examples/run_lm_finetuning.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (matej-svejda) [reply examples/run_lm_finetuning.py]: Hmm... are you sure we should divide twice by `args. gradient_accumulation_steps`?
train_batch_size is already getting divided by it on line 414 in run_classifier.py for example:
`args.train_batch_size = int(args.train_batch_size / args.gradient_accumulation_steps)`

[the remark this answers is not in the document]
Comment (donglixp) [reply examples/run_lm_finetuning.py]: The original version is correct. (The second time recovers the batch size without considering gradient_accumulation_steps.)

[the remark this answers is not in the document]
Comment (matej-svejda) [reply examples/run_lm_finetuning.py]: Yeah, you guys are right. I reverted the error I made here.

[the remark this answers is not in the document]
Comment (matej-svejda) [reply examples/run_lm_finetuning.py]: @thomwolf should `num_train_steps` be divided by `torch.distributed.get_world_size()` or only `t_total`? Because so far in all examples (except run_lm_finetuning.py) `t_total` already gets divided by `torch.distributed.get_world_size()` if `args.local_rank != -1`. I added that logic to run_lm_finetuning.py in my latest commit as well.

**Comment by thomwolf** [comment]
Thanks @matej-svejda, I was a bit late on this PR. I've made a small commit to make the notation more explicit (removed `t_total` which was mainly a duplicate of `num_train_steps` and renamed `num_train_steps` in a more explicit `num_train_optimization_steps`).
Merging this now
